test/integration/logic/TypeResolverTest.java

```diff
@@ -925,9 +970,33 @@ public void infer_key_attributes() {
 
         Map<String, Set<String>> expected = new HashMap<String, Set<String>>() {{
             put("$x", set("person"));
+            put("$_0", set("name"));
+            put("$_name", set("name"));
         }};
 
         assertEquals(expected, getHintMap(conjunction));
     }
 
+    @Test
+    public void role_labels_reduced_by_full_type_resolver() {
+        define_custom_schema("define" +
+                                     " person sub entity, plays partnership:partner;" +
+                                     " partnership sub relation, relates partner;" +
+                                     " business sub relation, relates partner;"
+        );
+
+        TypeResolver typeResolver = transaction.logic().typeResolver();
+        String queryString = "match (partner: $x);";
+        Conjunction conjunction = createConjunction(queryString);
+        typeResolver.resolveLabels(conjunction);
+
+        Set<String> expectedLabels = set("partnership:partner", "business:partner");
+
+        assertEquals(expectedLabels, getHintMap(conjunction).get("$_relation:partner"));
+
+        typeResolver.resolve(conjunction);
+        Set<String> expectedResolvedTypes = set("partnership:partner");
+
+        assertEquals(expectedResolvedTypes, getHintMap(conjunction).get("$_relation:partner"));
+    }
```

Comment:
@maxbaxt  this is for reduction of the set of roles. However I think there should also be a test that uses `match $role type relation:friend; ($role: $x) isa $y;` or the like.